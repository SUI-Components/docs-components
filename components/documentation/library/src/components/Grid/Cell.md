# Cell
> version undefined

div element css-grid styled in order to allocate elements in the grid disposed.

## Installation
`npm install @s-ui/documentation-library`

## Usage
```scss
// include Cell SASS module
@include '@s-ui/documentation-library/lib/components/Grid/Grid.scss'
```

```js
// import Cell DOC component
import { Cell } from '@s-ui/documentation-library/lib/components/Grid/Grid.js'
```

or

```scss
// include all SASS modules
@include '@s-ui/documentation-library/lib/index.scss'
```

```js
// import all DOC components you need
import { Cell } from '@s-ui/documentation-library'
```

## Prop-Types

| Name | Description | Values  | Default | Required |
| ---: |:---:| ---:| ---: |:---: |
| align |  | | |  [ ]  |
| offset | undefined | | |  [ ]  |
| span | undefined | | |  [ ]  |
| elementType | undefined | | |  [ ]  |
