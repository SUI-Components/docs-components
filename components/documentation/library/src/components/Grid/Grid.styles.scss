.sui-studio-doc-grid {
  display: grid;
  &-mode {
    &-light {
      color: var(--c-sui-studio-doc-base);
    }
    &-dark {
      color: var(--c-sui-studio-doc-neutral0);
    }
  }
  &-cell {
    &-ta {
      &-left {
        text-align: left;
      }
      &-right {
        text-align: right;
      }
      &-center {
        text-align: center;
      }
    }
  }
}
