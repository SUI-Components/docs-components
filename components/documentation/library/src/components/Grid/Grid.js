/* eslint react/prop-types: 0 */
import React, {forwardRef} from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import {withDocumentationProvider} from '../Base.core'

import './Grid.scss'

const range = (length, start = 0) => {
  return Array.from({length}, (_, i) => start + i)
}
const isNumeric = num => !isNaN(num)

export const ALIGN = {
  LEFT: 'left',
  RIGHT: 'right',
  CENTER: 'center'
}

/**
 * div element css-grid styled in order to allocate elements
 */
let Grid = forwardRef(
  (
    {
      children,
      className,
      gutter = 0,
      cols = 1,
      style,
      mode,
      elementType = 'div',
      ...props
    },
    forwardedRef
  ) => {
    const element = {type: elementType}
    const [gridRowGap, gridColumnGap] = `${gutter}`.split(',')
    return (
      <element.type
        {...props}
        style={{
          gridRowGap: isNumeric(gridRowGap) ? `${gridRowGap}px` : gridRowGap,
          gridColumnGap: isNumeric(gridColumnGap)
            ? `${gridColumnGap}px`
            : gridColumnGap,
          gridTemplateColumns: `repeat(${cols}, 1fr)`,
          ...style
        }}
        data-theme-mode={mode}
        className={cx('sui-studio-doc-grid', className)}
        ref={forwardedRef}
      >
        {children}
      </element.type>
    )
  }
)

Grid = withDocumentationProvider(Grid)
Grid.displayName = 'Grid'
Grid.propTypes = {}
Grid.defaultProps = {}

/**
 * div element css-grid styled in order to allocate elements in the grid disposed.
 */
let Cell = forwardRef(
  (
    {
      children,
      className,
      mode,
      offset = 0,
      span = 1,
      style,
      align,
      elementType = 'div',
      ...props
    },
    forwardedRef
  ) => {
    const element = {type: elementType}
    return (
      <>
        {range(offset).map(index => (
          <element.type
            className={cx(
              'sui-studio-doc-grid-cell',
              'sui-studio-doc-grid-cell-offset',
              className
            )}
            style={{gridColumn: 'auto / span 1'}}
            key={index}
          />
        ))}
        <element.type
          {...props}
          style={{
            gridColumn: `auto / span ${span}`,
            ...style
          }}
          data-theme-mode={mode}
          className={cx(
            'sui-studio-doc-grid-cell',
            {
              [`sui-studio-doc-grid-cell-ta-${align}`]: align
            },
            className
          )}
          ref={forwardedRef}
        >
          {children}
        </element.type>
      </>
    )
  }
)

Cell = withDocumentationProvider(Cell)

Cell.displayName = 'Cell'
Cell.propTypes = {
  align: PropTypes.oneOf(Object.values(ALIGN))
}
Cell.defaultProps = {}

export default Grid

export {Cell, Grid}
