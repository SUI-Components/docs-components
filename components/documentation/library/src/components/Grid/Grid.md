# Grid
> version undefined

div element css-grid styled in order to allocate elements

## Installation
`npm install @s-ui/documentation-library`

## Usage
```scss
// include Grid SASS module
@include '@s-ui/documentation-library/lib/components/Grid/Grid.scss'
```

```js
// import Grid DOC component
import { Grid } from '@s-ui/documentation-library/lib/components/Grid/Grid.js'
```

or

```scss
// include all SASS modules
@include '@s-ui/documentation-library/lib/index.scss'
```

```js
// import all DOC components you need
import { Grid } from '@s-ui/documentation-library'
```

## Prop-Types

| Name | Description | Values  | Default | Required |
| ---: |:---:| ---:| ---: |:---: |
| gutter | undefined | | |  [ ]  |
| cols | undefined | | |  [ ]  |
| elementType | undefined | | |  [ ]  |
