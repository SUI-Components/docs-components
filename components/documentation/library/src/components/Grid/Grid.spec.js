/* global expect */
import React from 'react'
import ReactDOM from 'react-dom'

import {Grid} from './Grid'

import setupEnvironment from '../../helpers/setupEnvironment'

describe('Grid', () => {
  describe('default', () => {
    const Component = Grid
    const setup = setupEnvironment(Component)

    test('should render without crashing', () => {
      // Given
      const props = {}

      // When
      const component = <Component {...props} />

      // Then
      const div = document.createElement('div')
      ReactDOM.render(component, div)
      ReactDOM.unmountComponentAtNode(div)
    })

    test('should NOT render null', () => {
      // Given
      const props = {}

      // When
      const {container} = setup(props)

      // Then
      expect(container.innerHTML).toBeString()
      expect(container.innerHTML).not.toHaveLength(0)
    })

    test('should have propTypes and defaultProps ', () => {
      // Given
      const component = Component

      // When

      // Then
      expect(component.hasOwnProperty('propTypes')).toBeTruthy()
      expect(component.hasOwnProperty('defaultProps')).toBeTruthy()
    })

    test('should have proper displayName attribute', () => {
      // Given
      // const props = {}

      // When
      const component = Component

      // Then
      expect(component.displayName).toBeString()
      expect(component.displayName).not.toHaveLength(0)
      expect(component.displayName).toMatchSnapshot()
    })
  })
})
