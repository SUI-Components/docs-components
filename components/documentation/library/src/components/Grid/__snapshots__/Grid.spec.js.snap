// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid default should have proper displayName attribute 1`] = `"Grid"`;
